auth/password_hash_test.go

```diff
@@ -64,3 +67,394 @@ func TestSetBcryptCost(t *testing.T) {
+	for i := 0; i < numPasswords; i++ {
+		password := []byte(fmt.Sprintf("%s%d", key, i))
+		hashAndPassword := hashAndPassword{
+			hash:     bcryptHash(b, password),
+			password: password,
+		}
+		hashAndPasswords[i] = hashAndPassword
+		if warmCache {
+			key := authKey(hashAndPassword.hash, hashAndPassword.password)
+			cache.Put(key)
+		}
+	}
+	return hashAndPasswords
+}
+
+func BenchmarkCompareHashAndPassword100PercentCacheHit(b *testing.B) {
+	const maxCacheSize = 100
+	benchmarks := []struct {
+		name  string
+		cache Cache
+	}{
+		{"RandReplKeyCache", NewRandReplKeyCache(maxCacheSize)},
+		{"NoReplKeyCache", NewNoReplKeyCache(maxCacheSize)},
+	}
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			b.ReportAllocs()
+			cache := bm.cache
+			defer cache.Purge()
+			testData := generateTestData(b, cache, true, "foo", maxCacheSize)
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				hashAndPassword := testData[i%maxCacheSize]
+				_ = compareHashAndPassword(cache, hashAndPassword.hash, hashAndPassword.password)
+			}
+		})
+	}
+}
+
+func BenchmarkCompareHashAndPasswordCacheMissAndFill(b *testing.B) {
+	const maxCacheSize = 100
+	benchmarks := []struct {
+		name  string
+		cache Cache
+	}{
+		{"RandReplKeyCache", NewRandReplKeyCache(maxCacheSize)},
+		{"NoReplKeyCache", NewNoReplKeyCache(maxCacheSize)},
+	}
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			b.ReportAllocs()
+			cache := bm.cache
+			defer cache.Purge()
+			testData := generateTestData(b, cache, false, "bar", maxCacheSize)
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				hashAndPassword := testData[i%maxCacheSize]
+				_ = compareHashAndPassword(cache, hashAndPassword.hash, hashAndPassword.password)
+			}
+		})
+	}
+}
+
+func BenchmarkCompareHashAndPassword100PercentIncorrect(b *testing.B) {
+	const maxCacheSize = 100
+	benchmarks := []struct {
+		name  string
+		cache Cache
+	}{
+		{"RandReplKeyCache", NewRandReplKeyCache(maxCacheSize)},
+		{"NoReplKeyCache", NewNoReplKeyCache(maxCacheSize)},
+	}
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			b.ReportAllocs()
+			cache := bm.cache
+			defer cache.Purge()
+			hash := bcryptHash(b, []byte("foo"))
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				_ = compareHashAndPassword(cache, hash, []byte("baz"))
+			}
+		})
+	}
+}
+
+func BenchmarkPutAndContains(b *testing.B) {
+	const (
+		maxCacheSize = 100
+		numKeys      = 200
+	)
+	benchmarks := []struct {
+		name  string
+		cache Cache
+	}{
+		{"RandReplKeyCache", NewRandReplKeyCache(maxCacheSize)},
+		{"NoReplKeyCache", NewNoReplKeyCache(maxCacheSize)},
+	}
+
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			b.ReportAllocs()
+			cache := bm.cache
+			defer cache.Purge()
+			testData := generateTestData(b, cache, true, "foo", numKeys)
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				hashAndPassword := testData[i%numKeys]
```

Comment:
As discussed, don't need to pregenerate keys.  Can just use "key" + random int between 1 and numKeys.

Reply:
Makes sense, this change has been added to the latest commit.